src/Api/include/Api/LockFreeApiEventProducer.h

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_API_LOCK_FREE_API_EVENT_PRODUCER_H_
+#define ORBIT_API_LOCK_FREE_API_EVENT_PRODUCER_H_
+
+#include "Api/EncodedEvent.h"
+#include "OrbitProducer/LockFreeBufferCaptureEventProducer.h"
+#include "ProducerSideChannel/ProducerSideChannel.h"
+
+namespace orbit_api {
+
+// This class is used to enqueue orbit_api::ApiEvent events from multiple threads and relay them to
+// OrbitService in the form of orbit_grpc_protos::ProducerCaptureEventCaptureEvent events.
+class LockFreeApiEventProducer
+    : public orbit_producer::LockFreeBufferCaptureEventProducer<orbit_api::ApiEvent> {
+ public:
+  LockFreeApiEventProducer() {
+    BuildAndStart(orbit_producer_side_channel::CreateProducerSideChannel());
+  }
+
+  ~LockFreeApiEventProducer() { ShutdownAndWait(); }
+
+ protected:
+  [[nodiscard]] virtual orbit_grpc_protos::ProducerCaptureEvent* TranslateIntermediateEvent(
+      orbit_api::ApiEvent&& raw_api_event, google::protobuf::Arena* arena) override {
+    auto* capture_event =
+        google::protobuf::Arena::CreateMessage<orbit_grpc_protos::ProducerCaptureEvent>(arena);
+    auto* api_event = capture_event->mutable_api_event();
+    api_event->set_timestamp_ns(raw_api_event.timestamp_ns);
+    api_event->set_pid(raw_api_event.pid);
+    api_event->set_tid(raw_api_event.tid);
+    api_event->set_r0(raw_api_event.encoded_event.args[0]);
```

Comment:
Are there any plans for the (near future) to move from raw data to some more structured data?

Reply:
There is significant performance overhead associated with "string" in protobufs as it doesn't use the Arena to allocate memory. That was verified through profiling. There is a bigger plan to maybe move away from protobufs altogether in favor of (maybe) flatbuffers. In that case, we would have proper structured data.